### PR TITLE
[Schools] Using k-means for creating districts

### PR DESCRIPTION
**Pull request by MathiasMehl**
- Move all school functions to schools.py
- Add function in utility.py to create clusters using k-means

### CODE REVIEW

**Comment by NicEastvillage** [comment]
> Move all school functions to schools.py

Don't be afraid to make a data class, maybe SchoolSettings, to pass all the arguments from main to `setup_schools`. I think it will be a much prettier solution than passing the `args` variable; or parsing the arguments again; or making `setup_schools` signature very long.

File: school.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        # Number of new schools to be placed
+        number_new_schools = num_schools_default - len(xml_schools.findall("school"))
+    else:
+        # Else place new number of schools as according to input
+        number_new_schools = school_count - len(xml_schools.findall("school"))
+
+    if number_new_schools == 0:
+        return
+    if number_new_schools < 0:
+        print(f"Warning: {school_count} schools was requested, but there are already {len(xml_schools)} defined")
```
Comment (falkecarlsen) [inline school.py]: We should use logging now. Remember to import it.

```suggestion
        logging.warning(f"{school_count} schools was requested, but there are already {len(xml_schools)} defined")
```

File: school.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        num_schools_default = math.ceil(int(inhabitants) * schools_per_1000_inhabitants / 1000)
+
+        # Number of new schools to be placed
+        number_new_schools = num_schools_default - len(xml_schools.findall("school"))
+    else:
+        # Else place new number of schools as according to input
+        number_new_schools = school_count - len(xml_schools.findall("school"))
+
+    if number_new_schools == 0:
+        return
```
Comment (falkecarlsen) [inline school.py]: We could log an info to the user that no new schools are being inserted.

File: utility.py [hunk 3]
```diff
@@ -38,6 +39,34 @@ def distance(pos1: Tuple[float, float], pos2: Tuple[float, float]):
...
 
 
+def k_means_clusters(net: sumolib.net.Net, num_clusters: int):
+    from perlin import get_edge_pair_centroid
+    edges = net.getEdges()
+
+    centroid_edges = [get_edge_pair_centroid(edge.getShape()) for edge in edges]
+    centroids = kmeans(centroid_edges, num_clusters, iter=25)
+
+    clusters = [[] for x in range(num_clusters)]
```
Comment (falkecarlsen) [inline utility.py]: Very minor thing, but I bet Pycharm is mad that `x` is unused - you can rename it to `_` and it'll shut up.

**Review by falkecarlsen — CHANGES_REQUESTED** [review]
Nice, looks good. Some minor comments. 

Please do a read-through and see if there's important places where logging is missing. See this comment for how to decide level: https://github.com/d610f20/random-activitygen-py/pull/24#issuecomment-610271526

File: school.py (hunk 1 above)
Comment (falkecarlsen) [reply school.py]: _it is imported, disregard_

File: school.py (hunk 2 above)
Comment (MathiasMehl) [reply school.py]: Changed in 48d911155f995f23c7a6848ff9bf09c9969c200d

File: school.py (hunk 1 above)
Comment (MathiasMehl) [reply school.py]: Committed in 48d911155f995f23c7a6848ff9bf09c9969c200d

File: randomActivityGen.py [hunk 4]
```diff
@@ -60,91 +59,108 @@
...
+        # Work hours are missing, so we add some default work hours
+        logging.info("Work hours are missing from statistics, adding a default configuration")
+        work_hours = ET.SubElement(city, "workHours")
+        ET.SubElement(work_hours, "opening", {"hour": "28800", "proportion": "70"})  # 70% at 8.00
+        ET.SubElement(work_hours, "opening", {"hour": "30600", "proportion": "30"})  # 30% at 8.30
+        ET.SubElement(work_hours, "closing", {"hour": "43200", "proportion": "10"})  # 10% at 12.00
+        ET.SubElement(work_hours, "closing", {"hour": "61200", "proportion": "30"})  # 30% at 17.00
+        ET.SubElement(work_hours, "closing", {"hour": "63000", "proportion": "60"})  # 60% at 17.30
+
+
```
Comment (falkecarlsen) [inline randomActivityGen.py]: These two functions are duplicate from `gates.py` and `utility`, respectively. Must've happened during merge from master.

**Review by falkecarlsen — CHANGES_REQUESTED** [review]

File: randomActivityGen.py (hunk 4 above)
Comment (MathiasMehl) [reply randomActivityGen.py]: Yes, should be fixed now. e54ab46a9faab7f044f3db77ad2a00e099f2a25d and 3676551f21fc8693b52647596a11c23aa8e9d75d

**Review by falkecarlsen — APPROVED** [review]
Nice :+1:

**Review by NicEastvillage — APPROVED** [review]

**Pull request merged**